### ISSUE TITLE
gollvm:  cross compiling 

### ISSUE THREAD

**Issue by wyw900807** [golang/go#42388]
Dear Sir,
      I want to compile a program in the x86_64 environment, this program can run in the aarch64 environment, and I try " go build -o output -gccgoflags="--target=aarch64-linux"", but it cannot work.

**Comment by ianlancetaylor** [comment]
CC @thanm 

I believe that GoLLVM does not not current support aarch64-linux anyhow.

**Comment by wyw900807** [comment]
Thank you for your reply, but the documents of gollvm writes that gollvm is supported for aarch64 linux.

Which architectures and operating systems are supported for gollvm?
Gollvm is currently supported only for x86_64 and aarch64 Linux"

**Comment by ianlancetaylor** [comment]
OK, thanks.

What is the exact output that you get from your `go build` command?

**Comment by wyw900807** [comment]
It outputs as follows:

/usr/local/bin/llvm-goc: unable to determine target CPU features for target aarch64-unknown-linux

**Comment by erifan** [comment]
This is caused by the inconsistency between the triple_ lists in ArchCpusAttrs.h and the normalized triple_ that we get through the --target option, and the root cause is clang. In fact, if you set --target=x86_64-linux, the same error will be reported. I can fix this problem.

**Comment by wyw900807** [comment]
Thank you very much!

**Comment by thanm** [comment]
Thanks @erifan. I took a look and I can reproduce as well.  I assume this will involve changing and/or rerunning tools/capture-fcn-attributes.go -- are you planning of sending a patch?

**Comment by erifan** [comment]
@thanm Yes, I plan to send a patch, but it will be next week, is that ok?

**Comment by thanm** [comment]
>Yes, I plan to send a patch, but it will be next week, is that ok?

Yup, that's fine. Thanks.

**Comment by advancedwebdeveloper** [comment]
@wyw900807 , there are some common files for both Clang and llvm-goc.
They define which CPU architectures are supported and how specific CPU features/instruction sets are related.
Some other files are responsible for the calling conventions, in LLVM.

As for now - you can do the following (this duplicates what Clang itself would do):

1. Create a C++ source code file named, say, llvm_cpu_features_investigation.cpp
2. Populate it with the following content (you could improve a bit, to format the output):

> \#include "llvm/ADT/StringRef.h"
> \#include "llvm/ADT/StringMap.h"
> \#include "llvm/MC/SubtargetFeature.h"
> \#include "llvm/Support/Host.h"
> 
> using namespace llvm;
> SubtargetFeatures Features1;
> 
> int main (int argc, char **argv)
> {
> sys::getHostCPUName();
> StringMap<bool> HostFeatures;
> if (sys::getHostCPUFeatures(HostFeatures))
>        for (auto &F : HostFeatures)
>          Features1.AddFeature(F.first(), F.second);
>    
>  
>  printf("test %s", Features1.getString().c_str());
> printf("\nsomething else\n");
> return 0;
> }
> 
> 

3. Compile it:
> $ clang++-11 -I/usr/include/llvm-c-11 -I/usr/include/llvm-11 -L/usr/lib/i386-linux-gnu/ -lLLVM-11 ~/workarea/llvm_cpu_features_investigation.cpp

4. Run your resulting executable:

> $ ./a.out
> test 
(...your CPU instruction sets are listed here...)
>
>something else
>

5. Run the following command:

> $ llc-11 --version
> LLVM (http://llvm.org/):
>   LLVM version 11.0.0
>   
>   Optimized build.
>   Default target:
(...that would be your current target, not for cross-compiling)
>   Host CPU: 
(...there would be your CPU model, as LLVM understands it...)

You shall see some CPU features, show during (4), with a "-" prefix. Those are non-supported features, for a specific CPU model (5). You can't expected that to be used during -mtune

Since you are cross-compiling - it would be required to get corresponding CPU features (4) and the CPU model(5) definitions, on Aarch64 (probably you would either run under qemu or would use a system-on-chip/micro-controller, for that purpose).
Please provide specific information about targeted Aarch64 hardware.

**Comment by advancedwebdeveloper** [comment]
Github Actions give access to a relic Clang 9 - so I used it for an example.


Here is the workflow configuration file:
> name: LLVM
> 
> on:
>   create:
>     branches: [ main ]
>   pull_request:
>     branches: [ main ]
> 
> jobs:
>   build:
>     runs-on: ubuntu-latest
> 
>     steps:
>       - uses: actions/checkout@v2
> 
>       - name: Get Clang's version
>         run: clang --version
> 
>       - name: Check our path
>         run: pwd
>       - name: Extracting llc's host target
>         run: llc-9 --version | grep CPU
>       - name: Compile test tool, for retrieveing CPU features
>         run: clang -I/usr/lib/llvm-9/include/ -I/usr/lib/llvm-9/include/llvm-c -L/usr/lib/llvm-9/lib -lLLVM-9 -lstdc++ $GITHUB_WORKSPACE/llvm_cpu_features_investigation.cpp
> 
>       - name: Run the CPU feature extractor
>         run: ./a.out

![Screenshot from 2020-11-08 05-11-12](https://user-images.githubusercontent.com/5404074/98456060-537b0100-2181-11eb-9eff-80b9eae32faf.png)

**Comment by wyw900807** [comment]
@advancedwebdeveloper  Thank you for your help, I will try it

**Comment by erifan** [comment]
Hi, @thanm I thought gollvm already supports cross compiling, but it doesn't seem to be. Because we did not generate standard libraries for multiple architectures, the corresponding libraries can't be found during cross compiling. Updating ArchCpusAttrs.h is easy, but adding support for cross compiling requires much more effort. It seems that gccgo does not currently support cross compiling ?
@wyw900807 Maybe you need to compile the gollvm toolchain on Linux aarch64, and then compile your program.

**Comment by wyw900807** [comment]
OK, Thank you!

**Comment by thanm** [comment]
@erifan yes, there is almost certainly a lot of work to do to set up Gollvm as a cross compiler with respect to the libraries. The first step is getting the compiler to work properly however. 

Gccgo does support cross compilation, but not using the same model as vlang-- with gccgo you need a separate installation and compiler binary per target (as opposed to the clang model in which a single compiler binary supports multiple targets).
